### PR TITLE
Add Runestone grading support

### LINKED ISSUE #56: [ENHANCEMENT] Runestone grading support
## Description
We need Runestone autograding + export to Canvas support.

### Issue
Attached is an example (with some info redacted) of a Runestone grade book exported as a CSV. On the admin end of Runestone, we could also export by assignment if that's easier on this end of things.

## Acceptance Criteria 
- [ ] Grades can be transferred from Runestone

[mines_csstem_Gradebook-1686351703163.csv](https://github.com/CSCI128/101GradingScript/files/11710565/mines_csstem_Gradebook-1686351703163.csv)

### PR DESCRIPTION
**Pull request by ezrichards**
Closes #56

### CODE REVIEW

**Comment by ezrichards** [comment]
Note that calculateLatePenalty is made for Gradescope and still outputs in a Gradescope.csv file, so we need to rework this infrastructure later, but it does export if you just generate that csv anyways.

File: test/Factories.py [hunk 1]
```diff
@@ -365,3 +365,74 @@ def writeOutGradescopeGrades(_fileName: str, _gradeSheet: list[str]):
...
+        gradeDist.random_state = numpyGenerator
+        gradeSheet = []
+
+        # three lines that are in the runestone format parser
+        gradeSheet.append("0, 0")
+        gradeSheet.append("0, 3") # total points
+        gradeSheet.append("0, 0")
+        for student in _studentRoster:
+            studentScore = gradeDist.rvs()
+            gradeSheet.append(f"{student['email']},{round(studentScore,2) * 10}%")
```
Comment (TriHardStudios) [inline test/Factories.py:406]: Is there a lateness in runestone? If so we should be generating it.

File: test/test_FileHelpers_csvLoaders.py [hunk 2]
```diff
@@ -24,6 +25,11 @@ def setUp(self):
 
         self.loadCSVSaved = csvLoaders.loadCSV
 
+        # Runestone gen:
+        Factories.generateStudentRunestoneGradesForAssignment(self.roster, "Week 1 Readings", 10)
+        self.assertTrue(os.path.exists(self.runestoneGradesheetLocation))
+        self.assertTrue(os.path.isfile(self.runestoneGradesheetLocation))
+
     def tearDown(self):
         shutil.rmtree("./gradescope")
```
Comment (TriHardStudios) [inline test/test_FileHelpers_csvLoaders.py:34]: Add runestone folder to the teardown

**Review by gregbell26 — CHANGES_REQUESTED** [review]
The code changes look good! 

I would like to see some changes to the factories and the comments need to be a bit more consistent (you refer to runestone as gradescope a bit).

File: test/Factories.py [hunk 3]
```diff
@@ -365,3 +365,74 @@ def writeOutGradescopeGrades(_fileName: str, _gradeSheet: list[str]):
...
+        random.seed(12345)
+
+        # On average students get an 75% - 80% with a standard deviation of 12% - 15%.
+        #  Using a 75% avg and a 15% sd for this
+        gradeDist: stats.truncnorm = generateTruncNorm(mean=_assignmentPoints * .75, sd=_assignmentPoints * .15,
+                                                       lower=0, upper=_assignmentPoints)
+
+        # Seed scipy for reproducibility
+        numpyGenerator = numpy_rand.Generator(numpy_rand.PCG64(12345))
+        gradeDist.random_state = numpyGenerator
```
[the remark this answers is not in the document]
Comment (ezrichards) [reply test/Factories.py:397]: Yeah, I just based this off the current Gradescope code. Not sure if it does. You may want to investigate this.

File: FileHelpers/csvLoaders.py [hunk 4]
```diff
@@ -117,3 +117,62 @@ def loadGradescope(_filename):
...
+    for i, row in runestoneDF.iterrows():
+        runestoneDF.at[i, 'E-mail'] = row['E-mail'].split('@')[0]
+
+        # derive actual score from total points and score percentage
+        score = (totalPoints / 100) * float(row[assignment].split("%")[0])
+        runestoneDF.at[i, assignment] = score
+
+        # add phony columns for gradesheet format
+        runestoneDF.at[i, 'lateness_comment'] = ''
+        runestoneDF.at[i, 'Status'] = ''
```
[the remark this answers is not in the document]
Comment (ezrichards) [reply FileHelpers/csvLoaders.py:171]: Sounds good. We should probably just convert it as soon as we load it into a DF.

File: test/Factories.py (hunk 1 above)
Comment (ezrichards) [reply test/Factories.py:402]: Not that I can tell after investigation on our current admin panel, but I'll reach out to the Runestone people.

File: test/Factories.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TriHardStudios) [reply test/Factories.py:393]: The way that your is implemented doesn't. The Gradescope one uses `random.uniform`.

**Review by gregbell26 — APPROVED** [review]

**Pull request merged**